File: frontend/pos-app/src/app/core/services/subscriptions.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '@env/environment';
import { AuthService } from '../auth/auth.service';

// ========== DTOs ==========

export interface SubscriptionPlan {
    id: string;
    nombre: string;
    descripcion: string;
    frecuencia: 'DIARIA' | 'SEMANAL' | 'QUINCENAL' | 'MENSUAL';
    precio: number;
    productos: string; // JSON string
    activo: boolean;
}

export interface Subscription {
    id: string;
    plan?: SubscriptionPlan;
    clienteNombre: string;
    clienteTelefono?: string;
    clienteEmail?: string;
    direccionEntrega: string;
    comuna?: string;
    notasEntrega?: string;
    estado: 'ACTIVA' | 'PAUSADA' | 'CANCELADA';
    proximaEntrega?: string;
    fechaInicio: string;
    fechaPausa?: string;
    fechaCancelacion?: string;
    totalEntregas: number;
}

export interface SubscriptionDelivery {
    id: string;
    subscription?: Subscription;
    fecha: string;
    horaProgramada?: string;
    horaEntrega?: string;
    estado: 'PENDIENTE' | 'EN_RUTA' | 'ENTREGADO' | 'FALLIDO' | 'REPROGRAMADO';
    direccion: string;
    notas?: string;
}

export interface SubscriptionStats {
    activeSubscriptions: number;
    monthlyRevenue: number;
    pendingDeliveries: number;
}

export interface CreateSubscriptionRequest {
    planId: string;
    clienteNombre: string;
    clienteTelefono?: string;
    direccionEntrega: string;
    comuna?: string;
    fechaInicio?: string;
}

@Injectable({
    providedIn: 'root'
})
export class SubscriptionsService {
    private baseUrl = environment.apiUrl || 'http://localhost:8080/api';

    constructor(
        private http: HttpClient,
        private authService: AuthService
    ) { }

    private getHeaders() {
        return {
            'X-Tenant-ID': this.authService.getTenantId() || '',
            'Authorization': `Bearer ${this.authService.getToken()}`
        };
    }

    // ========== PLANS ==========

    getPlans(activeOnly = true): Observable<SubscriptionPlan[]> {
        return this.http.get<SubscriptionPlan[]>(`${this.baseUrl}/subscriptions/plans`, {
            headers: this.getHeaders(),
            params: { activeOnly: activeOnly.toString() }
        });
    }

    // ========== SUBSCRIPTIONS ==========

    getSubscriptions(activeOnly = false): Observable<Subscription[]> {
        return this.http.get<Subscription[]>(`${this.baseUrl}/subscriptions`, {
            headers: this.getHeaders(),
            params: { activeOnly: activeOnly.toString() }
        });
    }

    getSubscription(id: string): Observable<Subscription> {
        return this.http.get<Subscription>(`${this.baseUrl}/subscriptions/${id}`, {
            headers: this.getHeaders()
        });
    }

    createSubscription(request: CreateSubscriptionRequest): Observable<Subscription> {
        return this.http.post<Subscription>(`${this.baseUrl}/subscriptions`, request, {
            headers: this.getHeaders()
        });
    }

    updateStatus(id: string, estado: string): Observable<Subscription> {
        return this.http.put<Subscription>(`${this.baseUrl}/subscriptions/${id}/status`, { estado }, {
            headers: this.getHeaders()
        });
    }

    // ========== DELIVERIES ==========

    getDeliveries(date?: string, pendingOnly = false): Observable<SubscriptionDelivery[]> {
        let params = new HttpParams();
        if (date) {
            params = params.set('date', date);
        }
        params = params.set('pendingOnly', pendingOnly.toString());

        return this.http.get<SubscriptionDelivery[]>(`${this.baseUrl}/subscriptions/deliveries`, {
            headers: this.getHeaders(),
            params
        });
    }

    updateDeliveryStatus(id: string, estado: string): Observable<SubscriptionDelivery> {
        return this.http.put<SubscriptionDelivery>(`${this.baseUrl}/subscriptions/deliveries/${id}/status`, { estado }, {
            headers: this.getHeaders()
        });
    }

    // ========== STATS ==========

    getStats(): Observable<SubscriptionStats> {
        return this.http.get<SubscriptionStats>(`${this.baseUrl}/subscriptions/stats`, {
            headers: this.getHeaders()
        });
    }
}

File: backend/operations-service/src/main/java/com/poscl/operations/application/service/SubscriptionsService.java
package com.poscl.operations.application.service;

import com.poscl.operations.domain.entity.Subscription;
import com.poscl.operations.domain.entity.SubscriptionDelivery;
import com.poscl.operations.domain.entity.SubscriptionPlan;
import com.poscl.operations.domain.repository.SubscriptionDeliveryRepository;
import com.poscl.operations.domain.repository.SubscriptionPlanRepository;
import com.poscl.operations.domain.repository.SubscriptionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class SubscriptionsService {

    private final SubscriptionRepository subscriptionRepository;
    private final SubscriptionPlanRepository planRepository;
    private final SubscriptionDeliveryRepository deliveryRepository;

    // ==================== PLANS ====================

    public List<SubscriptionPlan> getActivePlans(UUID tenantId) {
        return planRepository.findByTenantIdAndActivoOrderByPrecioAsc(tenantId, true);
    }

    public List<SubscriptionPlan> getAllPlans(UUID tenantId) {
        return planRepository.findByTenantIdOrderByNombreAsc(tenantId);
    }

    @Transactional
    public SubscriptionPlan createPlan(UUID tenantId, String nombre, String descripcion,
                                       String frecuencia, BigDecimal precio, String productos) {
        SubscriptionPlan plan = SubscriptionPlan.builder()
                .tenantId(tenantId)
                .nombre(nombre)
                .descripcion(descripcion)
                .frecuencia(frecuencia)
                .precio(precio)
                .productos(productos)
                .build();

        log.info("Creating subscription plan: {} for tenant: {}", nombre, tenantId);
        return planRepository.save(plan);
    }

    // ==================== SUBSCRIPTIONS ====================

    public List<Subscription> getSubscriptions(UUID tenantId) {
        return subscriptionRepository.findByTenantIdOrderByCreatedAtDesc(tenantId);
    }

    public List<Subscription> getActiveSubscriptions(UUID tenantId) {
        return subscriptionRepository.findByTenantIdAndEstadoOrderByCreatedAtDesc(tenantId, "ACTIVA");
    }

    public Optional<Subscription> getSubscriptionById(UUID id) {
        return subscriptionRepository.findById(id);
    }

    @Transactional
    public Subscription createSubscription(UUID tenantId, UUID planId, String clienteNombre,
                                           String clienteTelefono, String direccionEntrega,
                                           String comuna, LocalDate fechaInicio) {
        
        SubscriptionPlan plan = planRepository.findById(planId)
                .orElseThrow(() -> new IllegalArgumentException("Plan not found: " + planId));

        LocalDate proximaEntrega = calculateNextDelivery(fechaInicio, plan.getFrecuencia());

        Subscription subscription = Subscription.builder()
                .tenantId(tenantId)
                .plan(plan)
                .clienteNombre(clienteNombre)
                .clienteTelefono(clienteTelefono)
                .direccionEntrega(direccionEntrega)
                .comuna(comuna)
                .fechaInicio(fechaInicio)
                .proximaEntrega(proximaEntrega)
                .build();

        log.info("Creating subscription for {} with plan {}", clienteNombre, plan.getNombre());
        return subscriptionRepository.save(subscription);
    }

    @Transactional
    public Subscription updateStatus(UUID subscriptionId, String newStatus) {
        Subscription subscription = subscriptionRepository.findById(subscriptionId)
                .orElseThrow(() -> new IllegalArgumentException("Subscription not found: " + subscriptionId));

        switch (newStatus.toUpperCase()) {
            case "PAUSADA" -> subscription.pause();
            case "ACTIVA" -> subscription.resume();
            case "CANCELADA" -> subscription.cancel();
            default -> subscription.setEstado(newStatus);
        }

        log.info("Updated subscription {} status to {}", subscriptionId, newStatus);
        return subscriptionRepository.save(subscription);
    }

    // ==================== DELIVERIES ====================

    public List<SubscriptionDelivery> getDeliveriesForDate(LocalDate date) {
        return deliveryRepository.findByFechaOrderByHoraProgramadaAsc(date);
    }

    public List<SubscriptionDelivery> getPendingDeliveries(LocalDate date) {
        return deliveryRepository.findByFechaAndEstadoOrderByHoraProgramadaAsc(date, "PENDIENTE");
    }

    @Transactional
    public SubscriptionDelivery updateDeliveryStatus(UUID deliveryId, String newStatus) {
        SubscriptionDelivery delivery = deliveryRepository.findById(deliveryId)
                .orElseThrow(() -> new IllegalArgumentException("Delivery not found: " + deliveryId));

        switch (newStatus.toUpperCase()) {
            case "EN_RUTA" -> delivery.startDelivery();
            case "ENTREGADO" -> {
                delivery.markDelivered();
                // Update subscription delivery count
                Subscription subscription = delivery.getSubscription();
                subscription.setTotalEntregas(subscription.getTotalEntregas() + 1);
                subscriptionRepository.save(subscription);
            }
            case "FALLIDO" -> delivery.markFailed("Entrega fallida");
            default -> delivery.setEstado(newStatus);
        }

        log.info("Updated delivery {} status to {}", deliveryId, newStatus);
        return deliveryRepository.save(delivery);
    }

    // ==================== STATS ====================

    public SubscriptionStats getStats(UUID tenantId) {
        Long activeCount = subscriptionRepository.countByTenantIdAndEstado(tenantId, "ACTIVA");
        BigDecimal mrr = subscriptionRepository.calculateMonthlyRecurringRevenue(tenantId);
        Long pendingDeliveries = deliveryRepository.countByFechaAndEstado(LocalDate.now(), "PENDIENTE");

        return new SubscriptionStats(
                activeCount != null ? activeCount : 0,
                mrr != null ? mrr : BigDecimal.ZERO,
                pendingDeliveries != null ? pendingDeliveries : 0
        );
    }

    // ==================== HELPERS ====================

    private LocalDate calculateNextDelivery(LocalDate from, String frecuencia) {
        return switch (frecuencia.toUpperCase()) {
            case "DIARIA" -> from.plusDays(1);
            case "SEMANAL" -> from.plusWeeks(1);
            case "QUINCENAL" -> from.plusWeeks(2);
            case "MENSUAL" -> from.plusMonths(1);
            default -> from.plusDays(1);
        };
    }

    public record SubscriptionStats(long activeSubscriptions, BigDecimal monthlyRevenue, long pendingDeliveries) {}
}
